internal/security/secretstoreclient/roottoken.go

```diff
@@ -0,0 +1,176 @@
+		vc.logger.Error(fmt.Sprintf("failed to decode root token: %s", err.Error()))
+		return err
+	}
+
+	return nil
+}
+
+func (vc *vaultClient) rootTokenCancelPrevious() error {
+	_, err := vc.doRequest(commonRequestArgs{
+		AuthToken:            "",
+		Method:               http.MethodDelete,
+		Path:                 RootTokenControlAPI,
+		JSONObject:           nil,
+		BodyReader:           nil,
+		OperationDescription: "cancel previous root token generation",
+		ExpectedStatusCode:   http.StatusNoContent,
+		ResponseObject:       nil,
+	})
+	return err
+}
+
+func (vc *vaultClient) rootTokenStartGeneration(nonce *string, otp *string) error {
+	var response RootTokenControlResponse
+	_, err := vc.doRequest(commonRequestArgs{
+		AuthToken:            "",
+		Method:               http.MethodPut,
+		Path:                 RootTokenControlAPI,
+		JSONObject:           struct{}{},
+		BodyReader:           nil,
+		OperationDescription: "start root token generation",
+		ExpectedStatusCode:   http.StatusOK,
+		ResponseObject:       &response,
+	})
+	*nonce = response.Nonce
+	*otp = response.Otp
+	return err
+}
+
+func (vc *vaultClient) rootTokenSubmitKeys(vmkReader io.Reader, nonce string, encodedToken *string) error {
+	readCloser := fileioperformer.MakeReadCloser(vmkReader)
+	rawBytes, err := ioutil.ReadAll(readCloser)
+	defer readCloser.Close()
+	if err != nil {
+		msg := fmt.Sprintf("failed to read the Vault JSON response init file: %s", err.Error())
+		vc.logger.Error(msg)
+		return errors.New(msg)
+	}
+
+	initResp := InitResponse{}
+	if err = json.Unmarshal(rawBytes, &initResp); err != nil {
+		msg := fmt.Sprintf("failed to build the JSON structure from the init response body: %s", err.Error())
+		vc.logger.Error(msg)
+		return errors.New(msg)
+	}
+
+	for _, key := range initResp.KeysBase64 {
+		complete, err := vc.rootTokenSubmitKey(key, nonce, encodedToken)
+		if err != nil {
+			vc.logger.Error(fmt.Sprintf("root token retrieval aborted due to error: %s", err.Error()))
+			return err
+		} else if complete {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (vc *vaultClient) rootTokenSubmitKey(key string, nonce string, encodedToken *string) (bool, error) {
+	params := RootTokenRetrievalRequest{
+		Key:   key,
+		Nonce: nonce,
+	}
+	var response RootTokenRetrievalResponse
+	_, err := vc.doRequest(commonRequestArgs{
+		AuthToken:            "",
+		Method:               http.MethodPut,
+		Path:                 RootTokenRetrievalAPI,
+		JSONObject:           params,
+		BodyReader:           nil,
+		OperationDescription: "submit root token keyshare",
+		ExpectedStatusCode:   http.StatusOK,
+		ResponseObject:       &response,
+	})
+	if response.Complete {
+		*encodedToken = response.EncodedToken
+	}
+	return response.Complete, err
+}
+
+func (vc *vaultClient) rootTokenDecodeToken(encodedToken string, otp string, rootToken *string) error {
+	// otp is base62 ascii pad cast to raw bytes
```

Comment:
base64

Reply:
Believe it or not, this is not a typo.

Reply:
See https://www.vaultproject.io/api/system/generate-recovery-token.html

Reply:
It is good then